README.md
# [:octocat:TcStorage](https://github.com/TCCinTaiwan/TcStorage)
[![Version](https://img.shields.io/badge/lastest_version-0.1.4-blue.svg)](https://github.com/TCCinTaiwan/TcStorage/releases) [![Gitter](https://badges.gitter.im/TCCinTaiwan/TcStorage.svg)](https://gitter.im/TCCinTaiwan/TcStorage?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge)

## Contents[↶](#)
* **[Introduction](#introduction)**
* **[Installation](#installation)**
* **[ShortKey](#shortkey)**
* **[Todo](#todo)**
* **[History](#history)**
* **[Contributing](#contributing)**
* **[Languages](#languages)**

## Introduction[↶](#)
* Structure: PHP + MySQL + HTML5 + CSS3 + JS

這是以寫[:octocat:舊版](https://github.com/TCCinTaiwan/file-manager)的經驗為基礎，重構的新版本，多媒體參考[:octocat:TcTube](https://github.com/TCCinTaiwan/TcTube)部分內容，目前主要是在Chrome環境中執行，其他瀏覽器不一定支援。
![截圖-選取功能](screenshot/selection.gif)
![截圖-影片功能](screenshot/video.gif)
![截圖-動態歌詞功能](screenshot/multi-language-lyrics.gif)
![截圖-Ace Editor功能](screenshot/ace-editor.gif)

## Installation[↶](#)
1. Clone the repository.
    * Use Command(With submodule):
    ```bash
    cd C:/www/
    git clone --recursive https://github.com/TCCinTaiwan/TcStorage
    ```
    * Download Github zip:
        1. [ ![Download from Github](install/download-from-github.png)](https://github.com/TCCinTaiwan/TcStorage/archive/master.zip)
        2. upzip to your web server's document root.
        3. clone submodule repository.
        ```bash
        git submodule update --init --recursive
        ```

2. Running install page or step by step.
    * Install Page
    >The URL should like: http://localhost/TcStorage/install/index.php
    * Step by Step
        1. Create folders name "file/" and "logs/" .
        2. Import [TcStorage.sql](install/TcStorage.sql) at MySQL and change database name.
        3. Modify [connect.example.inc](install/connect.example.inc) to set the correct database connection information
        >Make sure the user has permission to view and modify the database

3. Ok, now you can open the browser using this system.
>The URL should like: http://localhost/TcStorage/index.php

## ShortKey[↶](#)
* <kbd>F2</kbd> 重新命名
* <kbd>Ctrl</kbd> + <kbd>A</kbd> 全選
* <kbd>Ctrl</kbd> + <kbd>I</kbd> 反選
* Ace Editor
    - <kbd>Ctrl</kbd> + <kbd>S</kbd> 儲存
    - <kbd>Esc</kbd> 顯示設定選單
    - <kbd>Ctrl</kbd> <kbd>Alt</kbd> <kbd>H</kbd> 列出快捷鍵
* Video/Audio

## Todo[↶](#)
* 多選 - shift
* 資料夾大小
* 檔案資訊
* 排列方式
* 帳戶
    - 分享網址
    - 權限
* 多語言
* 預覽
* 覆蓋選項
    - MD5檢查
* 使用提示

## History[↶](#)
Latest version is 0.1.4 , for detailed changelog, check [CHANGELOG.md](CHANGELOG.md).

## Contributing[↶](#)
1. Create an issue and describe your idea
2. [![GitHub forks](https://img.shields.io/github/forks/TCCinTaiwan/TcStorage.svg?style=social&label=Fork&maxAge=2592000)](https://github.com/TCCinTaiwan/TcStorage/network) it!
3. Create your feature branch: `git checkout -b my-new-feature`
4. Commit your changes: `git commit -m 'Add some feature'`
5. Push to the branch: `git push origin my-new-feature`
6. Submit a Pull Request

## Languages[↶](#)
1. Go to [PoEditor](https://poeditor.com/join/project/UQnirwwazV) join this project.
2. Translate!
3. Export your contributing language like main.po and main.mo file.
4. [![GitHub forks](https://img.shields.io/github/forks/TCCinTaiwan/TcStorage.svg?style=social&label=Fork&maxAge=2592000)](https://github.com/TCCinTaiwan/TcStorage/network) it!
5. Add files to locale/{en_US}/LC_MESSAGES depand on language.
6. Commit your changes: `git commit -m 'Update en_US locale files'`
7. Push to the branch: `git push origin my-new-feature`
8. Submit a Pull Request

CHANGELOG.md
# [:octocat:TcStorage](https://github.com/TCCinTaiwan/TcStorage)

## [↶Back to ReadMe](README.md)

## Contents[↶](#)
* **[0.1.4](#014)**
* **[0.1.3](#013)**
* **[0.1.2](#012)**
* **[0.1.1](#011)**
* **[0.1.0](#010)**

## 0.1.4[↶](#)
* 2019-09-01
    - [修改] 簡單調整 PHP 排版
    - [修正] 安裝檔案資料庫結構
    - [修改] AutoPay 提示
    - [修正] Ace 相對路徑錯誤問題
    - [修正] 加上時區
    - [修改] 更新 Ace 版本到 1.4.5

## 0.1.3[↶](#)
* 2018-06-17
    - [修正] 前端歌詞部分 JS 噴錯
    - [加入] 截圖改用 GIF 呈現
    - [修正] 右鍵在檔案標籤判定錯誤

## 0.1.2[↶](#)
* 2017-09-24
    - 整理.htaccess
    - 正名floatWindow底下物件(減少擴充功能衝突)
    - 解決iframe.contentDocument為null錯誤(像是PDF Viewer擴充功能會噴錯)
* 2017-09-25
    - 加入video/audio按一下播放暫停
    - 修正move為直接傳element
    - 重新命名檔案/資料夾功能簡單實現
    - 修正移動檔案錯誤
    - 流水號檔名判斷，假如已經是流水號檔名，則繼續編號
    - 開始寫檔頭/函數註解
    - 排除資料夾移動到自己
    - 移除與finishSelect功能衝突部分程式
    - 右鍵非選擇項目要先移除所選
    - F2快捷鍵
    - 拖曳複製(僅圖示無實現功能)
    - 麵包屑不顯示右鍵選單
* 2017-09-26
    - 加入LRC歌詞
    - 更新Google Analytics
    - $blockScrolling = Infinity;
    - 補上ACE漏掉的擴充功能引用(移到index.php)
    - LRC歌詞播完清空
    - 同路徑才使用(Query的先不管)
    - 刪除功能實現
    - Lyric搜尋文字先轉小寫
    - 複製功能草稿
    - Media播放完結束浮窗
* 2017-09-27
    - 把selectzone移到fileList底下
    - 解決滾輪偏移
    - 處理拖曳到自己顯示BUG
    - 設定"回上一層"不可選
    - 修改"回上一層"顏色
    - mouseDownInfo轉變成mouseInfo(加了滑鼠位置)
    - 選取框到邊邊滾動滾輪
    - 修復"回上一層"不一定存在BUG
    - 確保selectzone初始大小為一點
    - README.md資訊增加
    - 補上connect.example.inc
    - 加入全選跟反選
    - 快捷鍵：Ctrl+A全選
    - 整理raw.php讀檔
    - 啟用讀檔紀錄
    - 準備幾個快捷鍵
    - 快捷鍵：Ctrl+I反選
    - 選單按鍵部分靠右對齊
    - 加入快速安裝指南，並移除重複的connect.example.inc
* 2017-09-28
    - raw.php讀檔微調，Log檔位置修正
    - 整理安裝資料
    - 資料庫結構擴充(使用者/資料夾大小)
    - 更新資料庫匯入檔案
    - 麵包屑保持在頂端
    - 修正麵包屑調整後selectzone偏移
    - back、breadcrumbs不觸發selectzone
    - 建立多語言locale.inc
    - 調整路徑
    - 修改字體順位
    - 首頁多語系完善
* 2017-09-29
    - 移動History到CHANGLOG.md
    - 資料夾有無檔案判斷修正：加入子資料夾部分
    - 資料庫建立外來鍵
    - 因應資料庫改動[path_id]
    - MIME換成使用函數
    - 移除upload除錯用JSON
    - 修正$path_id為'0'資料庫應該是NULL問題
* 2017-10-01
    - 移動list與file功能到API資料夾
    - 測試 Travis Ci(尚未OK)
    - ZIP 加註解
    - ZIP 功能單一化，類似raw.php
    - 預覽ZIP前置作業
* 2017-10-02
    - 檔案清單增加更多時間資訊
* 2017-10-03
    - ZIP加到list
    - 選單加入複製貼上
    - 更新資料庫格式 補上資料夾時間
    - 加入預覽ZIP
    - 加入dirname()
    - ZIP檔案路徑CSS
    - 移除api/file.php
* 2017-10-04
    - 多層ZIP草稿
* 2017-10-11
    - 加入tooltip
    - 移除Travis Ci
    - 取消預覽ZIP
    - 加入js用的多國語言
    - 修改var成let(ES6)
    - 檔案資訊(ctime、size)
    - 影片LRC修正
    - lyric邊框

## 0.1.1[↶](#)
* 2017-02-25
    - 加入breadcrumbs預備元素
* 2017-02-26
    - 上一層資料夾改為雙擊
    - 修正修改檔案失敗
    - 加入breadcrumbs
    - breadcrumbs點擊
    - 修正圖片預覽
    - flv圖示
    - 右鍵部分功能實現
* 2017-02-27
    - 重命名微調
    - 實現選擇區域
    - 選擇區域切換視窗微調
    - 檔名重導向中文問題修正
    - 加入音頻預覽視覺化效果
    - 音頻視覺化微調
    - 擴充附檔名圖示

## 0.1.0[↶](#)
* 2017-01-25
    - 整理網頁，簡單重寫檔案管理
* 2017-01-26
    - 增加raw讀檔，並用.htaccess在網址中加入檔名
* 2017-01-27
    - 加入Ace
* 2017-01-28
    - raw合併下載
    - Ace加入標籤
* 2017-01-29
    - 命名為TcStorage
    - Ace改為多檔讀取
    - 改善IE 11相容性問題:xhr.response json vs object,函數預設值
* 2017-01-30
    - Ace開啟檔案編輯
    - 右鍵選單顯示並設定位置
    - 加入上傳檔案功能
    - 上傳Github
    - 調整上傳限制
    - Ace中文字型
    - 音樂影片開啟(含application/octet-stream)
    - 改善開啟多媒體檔案
* 2017-01-31
    - 加入指定mime
    - 讓raw媒體檔案Seekable
    - 修正編輯器跑版
    - Ace點擊url修正(加入相對位址)
    - Ace加入行列資訊
    - Ace加入語法選擇
    - Ace標籤調整顏色濾鏡(較明顯)
    - Ace加入重新讀取檔名選擇語法
    - 引用ajaxorg/ace
    - 將網址改為引用Repo
    - 移除原本ace-builds引用
    - 加入Font-Awesome引用
* 2017-02-01
    - Ace 網址自動格式化
    - Ace檔案存在修正
    - Ace.js整理
    - Ace檔案重新命名
    - Ace Base64編碼修正
    - 網站首頁選單加入Github連結
    - 修正檔名流水號錯誤問題
* 2017-02-02
    - Zip簡單讀取
    - Mime Type測試
    - 加入Mime Type清單更新
    - 建立函數引用檔
* 2017-02-03
    - 優化拖曳
    - 啟用Move
    - 加入可讀檔案大小
    - 優化拖曳流程
    - 整理事件
    - 優化UX
    - 加入多選(測試中)
* 2017-02-09
    - 安裝專案草案
    - 安裝目錄權限

api/list.php
<?php
/**
 *  API: 列出檔案/資料夾
 *
 * @version 0.1.3
 * @author TCC <john987john987@gmail.com>
 * @date 2019-09-01
 * @since 0.1.0 2017-09-29 TCC: 因應資料庫改動[path_id]
 * @since 0.1.0 2017-09-29 TCC: MIME換成使用函數
 * @since 0.1.1 2017-10-01 TCC: 移到API資料夾
 * @since 0.1.2 2017-10-02 TCC: 檔案清單增加更多時間資訊
 * @since 0.1.2 2017-10-03 TCC: ZIP加到list
 * @since 0.1.3 2019-09-01 TCC: 簡單調整 PHP 排版
 *
 * @param {Number} $_POST['id'] 目標資料夾ID或ZIP資料夾ID
 * @param {String} $_POST['zip'] ZIP檔案子路徑
 * @todo zip路徑大小寫忽略
 */
// if ($_SERVER['REQUEST_METHOD'] != "POST") { // DEBUG: 確保都是使用post
//     http_response_code(405); // Method not allowed
//     exit();
// }
include_once '../functions/connect.inc';
include_once '../functions/functions.inc';

header("Content-Type: application/json;charset=utf-8");

$zip = isset($_POST['zip']) ? $_POST['zip'] : (isset($_GET['zip']) ? $_GET['zip'] : null);
if ($zip === "") {
    $zip = ".";
} else if (!is_null($zip) && substr($zip, -1) === "/") {
    $zip = substr($zip, 0, -1);
}

$id = isset($_POST['id']) ? $_POST['id'] : (isset($_GET['id']) ? $_GET['id'] : "0");

$data = [
    "path_info" => [
        "id" => $id, // 預設為主目錄 DEBUG: $_GET是方便測試，請視情況拿掉
        "name" => "/", // 主目錄
        "root_id" => null, // 主目錄沒有上層
        "zip_path" => $zip, // 假如是壓縮檔會有這個路徑
        "full_path" => "",
        "breadcrumbs" => []
    ],
    "folders" => [],
    "files" => []
];

$query_id = $id === "0" ? 'IS NULL' : '= ' . $id;

if (!is_null($zip)) {
    $path_array = explode("/", rtrim($zip, "/"));
    $data['path_info']['name'] = end($path_array);

    for ($i = count($path_array) - 1; $i >= 0; $i--) {
        if ($path_array[$i] != ".") {
            array_push(
                $data['path_info']['breadcrumbs'],
                [
                    "name" => $path_array[$i],
                    "zip_path" => join("/", array_slice($path_array, 0, $i + 1))
                ]
            );
        }
    }

    $data['path_info']['full_path'] = $zip == "." ? "" : $zip;
    $sql = 'SELECT * FROM `files` WHERE `id` ' . $query_id . ';'; // TODO: ZIP
    $result = mysqli_query($conn, $sql);

    if ($row = mysqli_fetch_assoc($result)) {
        if ($data['path_info']['name'] == ".") {
            $data['path_info']['name'] = $row['name'];
        }

        $data['path_info']['full_path'] = $row['name'] . "/" . $data['path_info']['full_path'];

        array_push(
            $data['path_info']['breadcrumbs'],
            [
                "name" => $row['name'],
                "zip_path" => "."
            ]
        );

        $data['path_info']['root_id'] = $row['path_id'];
        $query_id = $row['path_id'] == "0" ? 'IS NULL' : '= ' . $row['path_id'];
    }
}

// 當前目錄資訊
$sql = 'SELECT * FROM `folders` WHERE `id` ' . $query_id . ';';
$result = mysqli_query($conn, $sql);

while ($row = mysqli_fetch_assoc($result)) { // 一直往上一層，並填充資料
    if (is_null($data['path_info']['root_id']) && is_null($zip)) { // 把第一層資料夾資訊填入
        $data['path_info']['name'] = $row['name'];
        $data['path_info']['root_id'] = is_null($row['path_id']) ? "0" : $row['path_id'];
    }

    $data['path_info']['full_path'] = $row['name'] . "/" . $data['path_info']['full_path'];

    array_push(
        $data['path_info']['breadcrumbs'],
        [
            "id" => $row['id'],
            "name" => $row['name']
        ]
    );

    $sql = 'SELECT * FROM `folders` WHERE `id` ' . (is_null($row['path_id']) ? 'IS NULL' : '= ' . $row['path_id']) . ';';

    $result = mysqli_query($conn, $sql);
}
// 加入主目錄資訊
$data['path_info']['full_path'] = "/" . $data['path_info']['full_path'];

array_push($data['path_info']['breadcrumbs'],
    [
        "id" => "0",
        "name" => "/"
    ]
);

$data['path_info']['breadcrumbs'] = array_reverse($data['path_info']['breadcrumbs']);

if (is_null($zip)) {
    // 子資料夾資訊(計算folder_count、file_count)
    $sql = 'SELECT *, (SELECT COUNT(*) FROM `folders` AS `count_folders` WHERE `count_folders`.`path_id` = `folders`.`id`) AS `folder_count`, (SELECT COUNT(*) FROM `files` AS `count_files` WHERE `count_files`.`path_id` = `folders`.`id`) AS `file_count` FROM `folders` WHERE `path_id` ' . $query_id . ';';

    $result = mysqli_query($conn, $sql);
    for ($index = 0; $row = mysqli_fetch_assoc($result); $index++) {
        $data['folders'][$index] = $row;
        $data['folders'][$index]['mtime'] = strtotime($data['folders'][$index]['mtime']);
        $data['folders'][$index]['ctime'] = strtotime($data['folders'][$index]['ctime']);
        $data['folders'][$index]['atime'] = strtotime($data['folders'][$index]['atime']);
        if (is_null($data['folders'][$index]['path_id'])) { // 將主目錄的路徑設為0
            $data['folders'][$index]['path_id'] = "0";
        }
    }

    // 子檔案資訊
    $sql = 'SELECT * FROM `files` WHERE `path_id` ' . $query_id . ' ORDER BY `name` ASC;';
    $result = mysqli_query($conn, $sql);
    for ($index = 0; $row = mysqli_fetch_assoc($result); $index++) {
        $data['files'][$index] = $row;
        $file_location = dirname(dirname(__FILE__)) . "\\files\\" . $row['id'];
        $data['files'][$index]['size'] = filesize($file_location);
        $data['files'][$index]['mtime'] = filemtime($file_location);
        $data['files'][$index]['ctime'] = filectime($file_location);
        $data['files'][$index]['atime'] = fileatime($file_location);

        if (is_null($data['files'][$index]['path_id'])) { // 將主目錄的路徑設為0
            $data['files'][$index]['path_id'] = "0";
        }
        if ($data['files'][$index]['mime'] == "") {
            $data['files'][$index]['mime'] = getFileMimeType($file_location, $data['files'][$index]['name']); // LINK: file.php:21 $info['mime'] = getFileMimeType($file_location, $info['name']);
        }

    }
} else {
    $file_location = dirname(dirname(__FILE__)) . "\\files\\" . $id;

    // zip讀取
    $zip_archive = new ZipArchive();
    $zip_archive->open($file_location);

    if ($zip_archive->numFiles > 0) {
        for($i = 0; $i < $zip_archive->numFiles; $i++) {
            $file_stat = $zip_archive->statIndex($i);
            // echo dirname($file_stat['name']) . "\n";
            $file_stat['path'] = $file_stat['name'];
            $file_stat['name'] = basename($file_stat['name']);
            if (dirname($file_stat['path']) == $zip) { // $zip_archive->getNameIndex($i)
                if (substr($file_stat['path'], -1) == "/") { // 資料夾
                    array_push(
                        $data['folders'],
                        $file_stat
                    );
                } else { // 檔案
                    $file_stat['mime'] = getFileMimeType("zip://" . $file_location . "#" . $file_stat['path'], $file_stat['name']);

                    array_push(
                        $data['files'],
                        $file_stat
                    );
                }
            }
        }
    }
}

echo json_encode($data);
